extractor/main_test.go

```diff
@@ -0,0 +1,412 @@
+		}
+	}
+}
+
+func TestProcessSourceFiles(t *testing.T) {
+	t.Parallel()
+
+	mockMsgs, mockMsgsAddPackage := generateMockMsgs(t)
+	sourceFiles := generateSourceFiles(t, mockMsgs)
+
+	var results []vm.MsgAddPackage
+	for _, sf := range sourceFiles {
+		res, err := processSourceFile(sf)
+		if err != nil {
+			t.Fatal(err)
+		}
+		results = append(results, res...)
+	}
+
+	sort.Slice(results, func(i, j int) bool {
+		return results[i].Package.Name < results[j].Package.Name
+	})
+	sort.Slice(mockMsgsAddPackage, func(i, j int) bool {
+		return mockMsgsAddPackage[i].Package.Name < mockMsgsAddPackage[j].Package.Name
+	})
+
+	if diff := deep.Equal(results, mockMsgsAddPackage); diff != nil {
+		fmt.Println(diff)
+	}
+}
+
+func TestWritePackageMetadata(t *testing.T) {
+	t.Parallel()
+
+	_, mockMsgsAddPackage := generateMockMsgs(t)
+
+	// Make temp dir
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, msg := range mockMsgsAddPackage {
+		md := metadataFromMsg(msg)
+
+		// Get output dir
+		outputDir := filepath.Join(tempDir, strings.TrimLeft(msg.Package.Path, "gno.land/"))
+
+		// Write dir before writing metadata
+		if dirWriteErr := os.MkdirAll(outputDir, os.ModePerm); dirWriteErr != nil {
+			t.Fatal(dirWriteErr)
+		}
+
+		// Write the metadata
+		err := writePackageMetadata(md, outputDir)
+		require.NoError(t, err)
+
+		// Read file
+		file, err := os.Open(filepath.Join(outputDir, packageMetadataFile))
+		require.NoError(t, err)
+
+		reader := bufio.NewReader(file)
+		var unmarshalledMetadata Metadata
+
+		raw, isPrefix, err := reader.ReadLine()
+		require.NoError(t, err)
+
+		if isPrefix {
+			t.Fatalf("Metadata longer then buffer max size at %s\n", outputDir)
+		}
+
+		err = json.Unmarshal(raw, &unmarshalledMetadata)
+		require.NoError(t, err)
+
+		t.Cleanup(func() {
+			err := os.RemoveAll(tempDir)
+			if err != nil {
+				fmt.Printf("could not clean up temp dir, %v", err)
+				return
+			}
+		})
+
+		require.Equal(t, md, unmarshalledMetadata)
+	}
+}
+
+func TestWritePackageFiles(t *testing.T) {
+	t.Parallel()
+
+	_, mockMsgsAddPackage := generateMockMsgs(t)
+
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, msg := range mockMsgsAddPackage {
+		// Get output dir
+		outputDir := filepath.Join(tempDir, strings.TrimLeft(msg.Package.Path, "gno.land/"))
+
+		// Write dir before writing metadata
+		if dirWriteErr := os.MkdirAll(outputDir, os.ModePerm); dirWriteErr != nil {
+			t.Fatal(dirWriteErr)
+		}
+
+		// Write the metadata
+		err := writePackageFiles(msg, outputDir)
+		require.NoError(t, err)
+
+		// Read & compare file
+		for _, f := range msg.Package.Files {
+			contents, err := os.ReadFile(filepath.Join(outputDir, f.Name))
+			require.NoError(t, err)
+			require.Equal(t, f.Body, string(contents))
+		}
+	}
+	t.Cleanup(func() {
+		err := os.RemoveAll(tempDir)
+		if err != nil {
+			fmt.Printf("could not clean up temp dir, %v", err)
+			return
+		}
+	})
```

Comment:
Should be right after you create the tempdir, not at the end of the method.
The problem is you set the cleanup at the end, which makes no sense -- if the test errors before (and it can), your stuck with leftover test directories

Reply:
Fixed in [25130b3](https://github.com/gnolang/tx-exports/pull/5/commits/25130b37b1d98224abe382e2cb9c33ee67009afe)

Reply:
See https://github.com/gnolang/tx-exports/pull/5#discussion_r1427808467

Reply:
Fixed as per [comment](https://github.com/gnolang/tx-exports/pull/5#discussion_r1438300888).

Reply:
Resolved in other conversations